# 09 — Tokens y límites de contexto

**Level 1 — LLM Engineering**

Medimos cuántos tokens consume cada texto con `tiktoken` (el tokenizador de OpenAI), vemos la relación palabras↔tokens en español, y truncamos textos para que quepan en la ventana de contexto.

In [1]:
import tiktoken


def contar_tokens(texto: str, modelo: str = "gpt-4o") -> int:
    """Cuenta los tokens de un texto usando el tokenizador del modelo."""
    enc = tiktoken.encoding_for_model(modelo)
    return len(enc.encode(texto))


def truncar(texto: str, max_tokens: int, modelo: str = "gpt-4o") -> str:
    """Trunca un texto a un maximo de tokens, conservando el principio."""
    enc = tiktoken.encoding_for_model(modelo)
    tokens = enc.encode(texto)
    if len(tokens) <= max_tokens:
        return texto
    return enc.decode(tokens[:max_tokens])

## Conteo de tokens

In [2]:
textos = [
    "hello world",
    "Hola mundo",
    "La inteligencia artificial transforma industrias enteras.",
    "supercalifragilisticoespialidoso",
]

print("=== CONTEO DE TOKENS ===")
for texto in textos:
    print(f"{contar_tokens(texto):>4} tokens | {texto}")

=== CONTEO DE TOKENS ===


   2 tokens | hello world
   2 tokens | Hola mundo
   8 tokens | La inteligencia artificial transforma industrias enteras.
  10 tokens | supercalifragilisticoespialidoso


## Relación palabras vs tokens (español)

In [3]:
parrafo = (
    "El aprendizaje automatico es una rama de la inteligencia artificial "
    "que permite a las computadoras aprender de los datos sin ser "
    "programadas explicitamente para cada tarea."
)
palabras = len(parrafo.split())
tokens = contar_tokens(parrafo)
print("=== RELACION PALABRAS vs TOKENS (español) ===")
print(f"palabras: {palabras}")
print(f"tokens:   {tokens}")
print(f"relacion: {tokens / palabras:.2f} tokens por palabra")

=== RELACION PALABRAS vs TOKENS (español) ===
palabras: 26
tokens:   31
relacion: 1.19 tokens por palabra


## Truncación

In [4]:
texto_largo = " ".join(["token de ejemplo numero"] * 100)
print("=== TRUNCACION ===")
print(f"texto original: {contar_tokens(texto_largo)} tokens")
cortado = truncar(texto_largo, 50)
print(f"truncado a 50:  {contar_tokens(cortado)} tokens")
print(f"final del texto truncado: ...{cortado[-30:]}")

=== TRUNCACION ===
texto original: 400 tokens
truncado a 50:  50 tokens
final del texto truncado: ...ken de ejemplo numero token de


## Conclusión

En español una palabra ≈ 1.19 tokens en promedio. El límite de contexto se gestiona recortando, no "resumiendo".